### PR TITLE
refactor MemStorage and adjust some tests

### PR DESCRIPTION
**Pull request by hicqu**
Extracted from #193 .
This PR makes `MemStorage` more like a real `Storage`. Some behaviors are changed:
1. `create_snapshot` don't need an `index` as parameter. It uses `committed` index directly.
2. `compact` now can compact all entries. Previously the last one will be always reserved.

### CODE REVIEW

**Comment by hicqu** [comment]
PTAL @Hoverbear @BusyJay @nolouch thank you!

File: src/raft_log.rs [hunk 1]
```diff
@@ -1321,19 +1321,19 @@ mod test {
                 1000,
                 vec![300, 500, 800, 900],
                 vec![700, 500, 200, 100],
-                true,
+                false,
```
Comment (Hoverbear) [inline src/raft_log.rs]: Why this change?

Comment (hicqu) [reply src/raft_log.rs]: Because I think this case shouldn't panic.

**Comment by hicqu** [comment]
I have no idea about why CI can't pass. Anyone can help me?

**Comment by hicqu** [comment]
Thanks for @BusyJay all tests are pass!

File: src/storage.rs [hunk 2]
```diff
@@ -275,6 +368,17 @@ impl MemStorage {
         }
     }
 
+    /// Create a `MemStorage` with a given `Config`.
+    pub fn initialize_with_config(&self, cfg: &Config) {
```
Comment (nolouch) [inline src/storage.rs]: This confused me.

File: src/storage.rs [hunk 3]
```diff
@@ -96,166 +113,242 @@ pub trait Storage {
...
+            let cs = self.raft_state.conf_states[i].clone();
+            snapshot.mut_metadata().set_conf_state(cs.conf_state);
         }
-        self.snapshot.set_data(data);
-        Ok(&self.snapshot)
+        snapshot
     }
 
-    /// Discards all log entries prior to compact_index.
+    /// Discards all log entries prior to compact_index (included).
```
Comment (nolouch) [inline src/storage.rs]: what does the included mean? such as [0, 1, 2, 3],  it means `compact(3)` will remains [3] but allow to compact(4) or `compact(3)` will remains empty.

File: src/storage.rs [hunk 4]
```diff
@@ -96,166 +113,242 @@ pub trait Storage {
...
+        Ok(())
+    }
+
+    /// Save raft logs and apply configuration chagnes in them.
+    pub fn append(&mut self, ents: &[Entry]) -> Result<()> {
+        if ents.is_empty() {
+            return Ok(());
+        }
+
+        // Gap is now allowed in raft logs.
```
[the remark this answers is not in the document]
Comment (hicqu) [reply src/storage.rs]: No, it should be `not allowed`. :(

File: src/storage.rs (hunk 2 above)
Comment (hicqu) [reply src/storage.rs]: Not used in this PR. I can remove it.

File: src/storage.rs (hunk 3 above)
Comment (hicqu) [reply src/storage.rs]: Stale comment. I will remove it.

File: src/storage.rs [hunk 5]
```diff
@@ -31,32 +31,44 @@
...
+use crate::util::limit_size;
+
+/// Used to track the history about configuration changes.
+#[derive(Clone, Debug, Default)]
+pub struct ConfStateWithIndex {
+    /// Target configuration state.
+    pub conf_state: ConfState,
+    /// Index of the entry.
+    pub index: u64,
+    /// Indicates it's a membership change or not.
```
Comment (Hoverbear) [inline src/storage.rs]:
```suggestion
    /// Indicates if the node is undergoing a membership change, or not.
```

File: src/storage.rs [hunk 6]
```diff
@@ -68,23 +80,28 @@ pub struct RaftState {
...
     fn term(&self, idx: u64) -> Result<u64>;
-    /// Returns the index of the first log entry that is
-    /// possible available via entries (older entries have been incorporated
-    /// into the latest snapshot; if storage only contains the dummy entry the
-    /// first log entry is not available).
+
+    /// Returns the index of the first log entry that is possible available via entries.
+    /// If the `Storage` is just initialized with a snapshot, `first_index` should return
+    /// the `snapshot index + 1`.
     fn first_index(&self) -> Result<u64>;
```
Comment (Hoverbear) [inline src/storage.rs]: Can we instruct what occurs when only the dummy entry is available?

File: src/storage.rs [hunk 7]
```diff
@@ -96,166 +113,234 @@ pub trait Storage {
...
+            if e.get_index() == index {
+                self.raft_state.hard_state.set_term(e.get_term());
+                self.raft_state.hard_state.set_commit(index);
+                break;
+            }
         }
+        Ok(())
+    }
+
+    /// Save raft logs and apply configuration chagnes in them.
```
Comment (Hoverbear) [inline src/storage.rs]:
```suggestion
    /// Save raft logs and apply configuration changes in them.
```

File: src/storage.rs [hunk 8]
```diff
@@ -96,166 +113,234 @@ pub trait Storage {
...
+    pub fn append(&mut self, ents: &[Entry]) -> Result<()> {
+        if ents.is_empty() {
+            return Ok(());
+        }
+
+        // Gap is not allowed in raft logs.
```
Comment (Hoverbear) [inline src/storage.rs]:
```suggestion
        // Gaps are not allowed in raft logs.
```

File: src/storage.rs [hunk 9]
```diff
@@ -528,47 +579,41 @@ mod test {
...
+        storage.wl().compact(4).unwrap();
+        assert_eq!(storage.first_index(), Ok(4));
     }
 
     #[test]
     fn test_storage_compact() {
         setup_for_test();
         let ents = vec![new_entry(3, 3), new_entry(4, 4), new_entry(5, 5)];
         let mut tests = vec![
-            (2, Err(RaftError::Store(StorageError::Compacted)), 3, 3, 3),
```
Comment (Hoverbear) [inline src/storage.rs]: So this no longer returns an error? Why?

File: tests/integration_cases/test_raw_node.rs [hunk 10]
```diff
@@ -57,22 +57,22 @@ fn conf_change(t: ConfChangeType, node_id: u64) -> ConfChange {
     cc
 }
 
-fn cmp_ready(
```
Comment (Hoverbear) [inline tests/integration_cases/test_raw_node.rs]: I don't understand why this change. This is less flexible (you can't assert `!cmp_ready(...)` etc).

Is it required? I would hate to see a `must_cmp_ready_false(...)` or something be added in the future :(

Comment (hicqu) [reply tests/integration_cases/test_raw_node.rs]: The reason about do this change is there are many assertions in the function, and they are chained with `&&`. So if one condition fails, we can't know which one it is. But if we do `assert` one by one in the function, we can know which condition fails easily.

File: src/storage.rs (hunk 9 above)
Comment (hicqu) [reply src/storage.rs]: I think in this case an error is unnecessary. Please consider this case:
```rust
storage.compact(100);
storage.compact(80);
```
Firstly we compact to 100, and then compact to 80. Previously it will return an Error, but I think we can just do nothing internally, and return `Ok`. What do you think?

File: src/storage.rs (hunk 6 above)
Comment (hicqu) [reply src/storage.rs]: Here I think we don't need to mention `dummy entry` because it's public document of `Storage`. Dummy entry is just a implementation detail of `MemStorage`.

Comment (Hoverbear) [reply src/storage.rs]: OK.

File: src/storage.rs (hunk 9 above)
Comment (Hoverbear) [reply src/storage.rs]: I think just no-oping it is fine. :)

File: tests/integration_cases/test_raw_node.rs (hunk 10 above)
Comment (Hoverbear) [reply tests/integration_cases/test_raw_node.rs]: Ok, let's hope if we need such a functionality in the future we will be clever and add a boolean flag instead of re-implement :)

File: src/storage.rs [hunk 11]
```diff
@@ -31,32 +31,44 @@
...
+    /// History configuration states and their indices. Not all history need to be reserved, the
+    /// `Storage` instance need only to ensure the first entry's index is less than applied index
+    /// so that when the peer sends snapshot to followers, it can choose a correct conf state.
+    ///
+    /// Entries in the field should be sorted by their indices.
+    pub conf_states: Vec<ConfStateWithIndex>,
+}
+
+impl RaftState {
+    /// Indicates the `RaftState` is initialized or not.
```
Comment (Hoverbear) [inline src/storage.rs]:
```suggestion
    /// Indicates if the `RaftState` is initialized or not.
```

File: src/storage.rs [hunk 12]
```diff
@@ -96,166 +112,241 @@ pub trait Storage {
...
     fn default() -> MemStorageCore {
         MemStorageCore {
+            raft_state: Default::default(),
             // When starting from scratch populate the list with a dummy entry at term zero.
+            // The dummy entry is used to ensure `entries` never be empty, which can simplify
+            // many logics.
             entries: vec![Entry::new()],
-            hard_state: HardState::new(),
-            snapshot: Snapshot::new(),
+            // Every time applys a snapshot to the storage, the metadata will be stored here,
```
Comment (Hoverbear) [inline src/storage.rs]:
```suggestion
            // Every time a snapshot is applied to the storage, the metadata will be stored here,
```

File: src/storage.rs [hunk 13]
```diff
@@ -96,166 +112,241 @@ pub trait Storage {
...
-            )
+
+        if compact_index > self.inner_last_index() + 1 {
+            return Err(Error::Store(StorageError::Unavailable));
         }
 
-        let i = (compact_index - offset) as usize;
-        let entries = self.entries.drain(i..).collect();
-        self.entries = entries;
+        let mut offset = compact_index - self.inner_first_index();
```
Comment (nolouch) [inline src/storage.rs]: maybe directly `offset =  compact_index - self.entries[0].get_index ` ?

**Comment by hicqu** [comment]
PTAL @nolouch @Hoverbear thank you!

File: src/storage.rs [hunk 14]
```diff
@@ -96,166 +112,237 @@ pub trait Storage {
...
+            // Every time a snapshot is applied to the storage, the metadata will be stored here,
+            // and a dummy entry will be pushed to `entries`.
+            snapshot_metadata: Default::default(),
         }
     }
 }
 
 impl MemStorageCore {
+    /// Append a configuration state. Only for tests.
+    pub fn append_conf_state(&mut self, cs: ConfStateWithIndex) {
```
Comment (BusyJay) [inline src/storage.rs]: Seems unrelated to this PR?

Comment (hicqu) [reply src/storage.rs]: It's used in `test_membership_change.rs`, for persisting configuration before power cycle.

File: src/storage.rs [hunk 15]
```diff
@@ -291,64 +356,42 @@ impl MemStorage {
...
         let max_size = max_size.into();
         let core = self.rl();
-        let offset = core.entries[0].get_index();
-        if low <= offset {
+        if low < core.inner_first_index() {
             return Err(Error::Store(StorageError::Compacted));
         }
 
         if high > core.inner_last_index() + 1 {
-            panic!(
```
Comment (BusyJay) [inline src/storage.rs]: I prefer panicing here as it's unreachable. Reach this line means either storage or raft is not implemented correctly, return `Unavailable` is misleading.

File: src/storage.rs [hunk 16]
```diff
@@ -126,136 +154,173 @@ impl MemStorageCore {
...
+            .position(|e| e.get_index() == first_idx)
+            .unwrap();
+        ents[i..].iter().for_each(|e| self.entries.push(e.clone()));
+        Ok(())
+    }
+
+    fn inner_first_index(&self) -> u64 {
+        let first_index = self.entries[0].get_index();
+        if self.snapshot_metadata.get_index() == first_index {
+            // The first entry's index equals to the snapshot index means the entry is "dummy"
```
Comment (BusyJay) [inline src/storage.rs]: How about removing the dummy entry completely?

File: src/storage.rs [hunk 17]
```diff
@@ -126,136 +154,173 @@ impl MemStorageCore {
...
-            self.snapshot
-                .mut_metadata()
-                .set_pending_membership_change_index(idx);
+            self.raft_state.pending_conf_state = Some(cs);
+            self.raft_state.pending_conf_state_start_index = Some(idx);
+        }
+    }
+
+    /// Save raft logs and apply configuration changes in them.
+    pub fn append(&mut self, ents: &[Entry]) -> Result<()> {
```
Comment (BusyJay) [inline src/storage.rs]: Why reorder the function? It is not easy to review.

File: src/storage.rs (hunk 16 above)
Comment (hicqu) [reply src/storage.rs]: I have tried. It proves that we need to set `snapshot_metadata` to a correct value in some cases. To avoid change those test cases, I prefer to keep it.

File: src/storage.rs (hunk 15 above)
Comment (hicqu) [reply src/storage.rs]: Here I think return an Error is better because `entries` is a public API. Users may use the interface in a wrong way. For that maybe an Error is better than panic.

Comment (BusyJay) [reply src/storage.rs]: How about documenting it clearly in the trait method? Correctness is more important.

Comment (hicqu) [reply src/storage.rs]: Done. PTAL thanks.

File: src/storage.rs [hunk 18]
```diff
@@ -68,23 +68,31 @@ pub struct RaftState {
...
     fn initial_state(&self) -> Result<RaftState>;
+
     /// Returns a slice of log entries in the range `[low, high)`.
     /// max_size limits the total size of the log entries returned if not `None`, however
     /// the slice of entries returned will always have length at least 1 if entries are
     /// found in the range.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `high` is higher than `Storage::last_index(self)`.
```
Comment (BusyJay) [inline src/storage.rs]: I think it should be higher than `storage::last_index(self) + 1`?

File: src/storage.rs [hunk 19]
```diff
@@ -96,28 +104,52 @@ pub trait Storage {
...
+    pub fn hard_state(&self) -> &HardState {
+        &self.raft_state.hard_state
+    }
+
+    /// Apply to an index so that we can crate a latest snapshot from the storage.
+    pub fn apply_to(&mut self, index: u64) -> Result<()> {
+        if index < self.inner_first_index() || index > self.inner_last_index() {
+            return Err(Error::Store(StorageError::Unavailable));
+        }
+        for e in &self.entries {
```
Comment (BusyJay) [inline src/storage.rs]: I think you can jump to the index directly instead of iterating.

File: src/storage.rs [hunk 20]
```diff
@@ -126,136 +158,173 @@ impl MemStorageCore {
...
+
+        if compact_index > self.inner_last_index() + 1 {
+            return Err(Error::Store(StorageError::Unavailable));
         }
 
-        let i = (compact_index - offset) as usize;
-        let entries = self.entries.drain(i..).collect();
-        self.entries = entries;
+        let offset = compact_index - self.entries[0].get_index();
+        self.entries = self.entries[offset as usize..].to_vec();
```
Comment (BusyJay) [inline src/storage.rs]: How about `self.entries.drain(..offset)`?

File: src/storage.rs [hunk 21]
```diff
@@ -96,28 +108,46 @@ pub trait Storage {
...
+
+    /// Get the hard state.
+    pub fn hard_state(&self) -> &HardState {
+        &self.raft_state.hard_state
+    }
+
+    /// Apply to an index so that we can crate a latest snapshot from the storage.
+    pub fn apply_to(&mut self, index: u64) -> Result<()> {
+        if index < self.inner_first_index() || index > self.inner_last_index() {
+            return Err(Error::Store(StorageError::Unavailable));
```
Comment (BusyJay) [inline src/storage.rs]: I think it should panic here.

File: src/storage.rs [hunk 22]
```diff
@@ -96,28 +108,46 @@ pub trait Storage {
...
+        }
+        let diff = (index - self.entries[0].get_index()) as usize;
```
Comment (BusyJay) [inline src/storage.rs]: What if it's called after applying a snapshot?

File: src/storage.rs [hunk 23]
```diff
@@ -96,28 +108,46 @@ pub trait Storage {
...
+        self.raft_state.hard_state.set_commit(index);
```
Comment (BusyJay) [inline src/storage.rs]: Why change commit?

File: src/storage.rs [hunk 24]
```diff
@@ -126,136 +156,163 @@ impl MemStorageCore {
...
-        let mut e = Entry::new();
-        e.set_term(snapshot.get_metadata().get_term());
-        e.set_index(snapshot.get_metadata().get_index());
-        self.entries = vec![e];
-        self.snapshot = snapshot;
+        self.snapshot_metadata = meta.clone();
+
+        // Update conf states.
+        self.raft_state.hard_state.set_term(term);
+        self.raft_state.hard_state.set_commit(index);
```
Comment (BusyJay) [inline src/storage.rs]: I think you need to clear `self.entries`.

File: src/storage.rs (hunk 23 above)
Comment (hicqu) [reply src/storage.rs]: I think it's better to chagne the function name to `commit_to`. What do you think?

File: src/storage.rs [hunk 25]
```diff
@@ -126,136 +164,164 @@ impl MemStorageCore {
...
-        } else if self.entries.len() as u64 == offset {
-            self.entries.extend_from_slice(te);
-        } else {
+        if self.inner_first_index() > ents[0].get_index() {
             panic!(
-                "missing log entry [last: {}, append at: {}]",
-                self.inner_last_index(),
-                te[0].get_index()
-            )
+                "overwrite compacted raft logs, compacted: {}, append: {}",
```
Comment (BusyJay) [inline src/storage.rs]: What if messages are reordered?

File: src/storage.rs [hunk 26]
```diff
@@ -126,136 +164,164 @@ impl MemStorageCore {
...
+        pending_membership_change: Option<ConfChange>,
+    ) -> Result<()> {
+        if idx <= self.snapshot_metadata.get_index() {
+            return Err(Error::Store(StorageError::SnapshotOutOfDate));
+        }
+        if idx > self.inner_last_index() {
+            return Err(Error::Store(StorageError::Unavailable));
+        }
+        for e in &self.entries {
+            if e.get_index() != idx {
```
Comment (BusyJay) [inline src/storage.rs]: Loop is not necessary.

File: src/storage.rs [hunk 27]
```diff
@@ -126,136 +164,164 @@ impl MemStorageCore {
...
+        ents.iter().for_each(|e| self.entries.push(e.clone()));
+        Ok(())
+    }
+
+    /// Initialize a snapshot in `MemStorageCore` with given index. Only used for tests.
+    pub fn create_snapshot(
+        &mut self,
+        idx: u64,
+        cs: Option<ConfState>,
+        pending_membership_change: Option<ConfChange>,
```
Comment (BusyJay) [inline src/storage.rs]: As a real storage, data is necessary.

File: src/storage.rs (hunk 25 above)
Comment (hicqu) [reply src/storage.rs]: You mean `ents` is not ordered? A correct Raft ensures it's ordered.

Comment (hicqu) [reply src/storage.rs]: If you mean some raft messages are delayed in network, It's Raft's responsibility to truncate redundant entries.

File: src/storage.rs (hunk 27 above)
Comment (hicqu) [reply src/storage.rs]: Emmm, key-value pairs (or other something after applied) are not stored in the storage. So I think it's weird to introduce `data` here. If users want to write a storage with real data stored, they can embed `MemStorage` into their own storage. What do you think?

File: src/storage.rs (hunk 25 above)
Comment (hicqu) [reply src/storage.rs]: Fixed.

**Comment by hicqu** [comment]
PTAL again @BusyJay thank you!

File: src/storage.rs [hunk 28]
```diff
@@ -269,6 +269,10 @@ impl MemStorageCore {
     }
 
     /// Append the new entries to storage.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `ents` is not expected.
```
Comment (BusyJay) [inline src/storage.rs]: What does "expected" mean?
